File: 48. SummaryRanges.java
import java.util.ArrayList;
import java.util.List;

public class SummaryRanges {

    public static List<String> summaryRanges(int[] nums) {
        List<String> out = new ArrayList<>();
        int i = 0, n = nums.length;
        while (i < n) {
            int start = nums[i];
            while (i + 1 < n && nums[i + 1] == nums[i] + 1) i++;
            if (start == nums[i]) out.add(String.valueOf(start));
            else out.add(start + "->" + nums[i]);
            i++;
        }
        return out;
    }

    public static void main(String[] args) {
        System.out.println(summaryRanges(new int[]{0, 1, 2, 4, 5, 7}));    // [0->2, 4->5, 7]
        System.out.println(summaryRanges(new int[]{0, 2, 3, 4, 6, 8, 9})); // [0, 2->4, 6, 8->9]
        System.out.println(summaryRanges(new int[]{}));                    // []
    }
}

File: 48. SummaryRanges.c
#include <stdio.h>

void summaryRanges(int *nums, int n) {
    int i = 0;
    printf("[");
    int first = 1;
    while (i < n) {
        int start = nums[i];
        while (i + 1 < n && nums[i + 1] == nums[i] + 1) i++;
        if (!first) printf(", ");
        first = 0;
        if (start == nums[i]) printf("%d", start);
        else printf("%d->%d", start, nums[i]);
        i++;
    }
    printf("]\n");
}

int main(void) {
    int a[] = {0, 1, 2, 4, 5, 7};
    summaryRanges(a, 6);    // [0->2, 4->5, 7]

    int b[] = {0, 2, 3, 4, 6, 8, 9};
    summaryRanges(b, 7);    // [0, 2->4, 6, 8->9]

    summaryRanges(NULL, 0); // []

    return 0;
}
